File: 2021/08/01/has_cycle.py
"""
https://leetcode.com/problems/linked-list-cycle
Author: Andrew Jarombek
Date: 8/1/2021
"""


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


def has_cycle(head: ListNode) -> bool:
    items = set()

    prev_id = None

    while head is not None:
        current_id = id(head)

        if current_id == prev_id:
            return False

        if current_id in items:
            return True

        items.add(id(head))
        head = head.next

    return False

File: 2022/02/17/c/hasCycle.c
// Program: https://leetcode.com/problems/linked-list-cycle/
// Author: Andrew Jarombek
// Date: 2/17/2022

#include <stdbool.h>
#include <stdlib.h>
#include <assert.h>

struct ListNode {
    int val;
    struct ListNode *next;
};

bool hasCycle(struct ListNode *head) {
    struct ListNode* node = head;
    int index = 0;

    while (node != NULL) {
        struct ListNode* prev = head;
        for (int i = 0; i < index; i++) {
            if (prev == node) {
                return true;
            }

            prev = prev->next;
        }

        index++;
        node = node->next;
    }

    return false;
}

int main() {
    struct ListNode node2 = { .val = 2, .next = NULL };
    struct ListNode node1 = {1, &node2 };

    assert(!hasCycle(&node1));

    node2.next = &node1;
    assert(hasCycle(&node1));
}
